Firestore/Source/Remote/FSTDatastore.m

```diff
@@ -585,6 +695,32 @@ - (void)inhibitBackoff {
   [self.backoff reset];
 }
 
+/** Called by the idle timer when the stream should close due to inactivity. */
+- (void)handleIdleCloseTimer {
+  [self.workerDispatchQueue verifyIsCurrentQueue];
+  if (self.state == FSTStreamStateOpen && [self isIdle]) {
+    // When timing out an idle stream there's no reason to force the stream into backoff when
+    // it restarts so set the stream state to Initial instead of Error.
+    [self close:FSTStreamStateInitial error:nil];
+  }
+}
+
+- (void)markIdle {
+  [self.workerDispatchQueue verifyIsCurrentQueue];
+  if (self.state == FSTStreamStateOpen) {
+    self.idle = YES;
+    [self.workerDispatchQueue dispatchAsyncAllowingSameQueue:^() {
+      [self handleIdleCloseTimer];
+    }
+                                                       after:kIdleTimeout];
```

Comment:
I'm not wild about this API. It's conflating two separate concerns: which queue to dispatch on and when.

This should be more like the backoff implementation: the dispatch queue should be wrapped by the idle timer, and we can supply an alternate implementation of the idle timer for tests. The idle timer could take the dispatch queue and idle timeout as initializer parameters

Reply:
Okay, we discussed offline, and this does make sense because we're modeling `dispatch_after` on iOS and a `ScheduledExecutorService` on Android.

I'm still not crazy about how this API ends up reading (especially because of the way clang format munges it). Maybe if we reversed the argument order it wouldn't look so nuts.

How about `dispatchAfter:block:`?

Reply:
Renamed to `dispatchAfterDelay:block:` and got rid of "allowsamequeue" as well.